src/data-list/data-list.service.ts

```diff
@@ -0,0 +1,276 @@
+        $selected?: boolean;
+    }
+
+    export interface ListConfig {
+        url: string;
+        id?: string;
+        paging?: boolean;
+        responseListName?: string;
+        responseCountName?: string;
+    }
+
+    export class DataList<T extends ListRow> {
+
+        id: string;
+        rows: T[];
+        url: string;
+        count: number;
+        data: any;
+        filter: any;
+        paging: boolean;
+        areFiltersSet: boolean;
+        forceListUpdate: boolean;
+        selectedAllPages: boolean;
+        isLoading: boolean;
+        responseListName: string;
+        responseCountName: string;
+
+        onListResponseSuccess: any;
+        onListResponseError: any;
+
+        initFilterFunc: Function;
+
+        constructor(config: ListConfig, private $http: ng.IHttpService, private $location: ng.ILocationService) {
+            this.id = config.id;
+            this.rows = [];
+            this.count = 0;
+            this.data = null;   // will be assigned raw last response
+            this.filter = { skip: 0, limit: 25 };
+            this.paging = config.paging;
+            this.areFiltersSet = false;
+            this.selectedAllPages = false;
+            this.isLoading = false;
+            this.url = config.url;
+            this.responseListName = config.responseListName || 'items';
+            this.responseCountName = config.responseCountName || 'count';
+            this.initFilterFunc = () => { return {}; };
+        }
+
+        onSuccess(callback: (list: T[], count: number) => void) {
+            this.onListResponseSuccess = callback;
+        }
+
+        onError(callback: (data: any) => void) {
+            this.onListResponseError = callback;
+        }
+
+        setFilter(filterFunc) {
+            this.initFilterFunc = filterFunc;
+            var filter = this.initFilterFunc();
+            if (typeof (filter) !== 'object') {
+                throw new Error('initFilterFunc should return object with filter properties');
+            }
+
+            angular.extend(this.filter, filter);
+
+            this.areFiltersSet = this.loadLocationParams(this.filter) > 0;
+        }
+
+        submitFilter() {
+            this.setLocationParams(this.filter);
+            return this.updateList();
+        }
+
+        resetFilter() {
+            this.filter = this.initFilterFunc();
+            this.setLocationParams(this.filter);
+            this.updateList();
+        }
+
+        updateList() {
+            const self = this;
+            this.isLoading = true;
+
+            if (typeof (this.filter.limit) === 'undefined') {
+                this.filter.limit = 25;
+            }
+
+            if (typeof (this.filter.skip) === 'undefined') {
+                this.filter.skip = 0;
+            }
+
+            const config = {
+                params: this.filter
+            };
+
+            return this.$http.get(this.url, config)
+                .success(data => {
```

Comment:
It is much easier to maintain code without these custom $http callbacks. Tests are easier to write with plain `.then`.